protocols/luci-proto-bonding/htdocs/luci-static/resources/protocol/bonding.js

```diff
@@ -0,0 +1,442 @@
+			var NotUsable = NonUsableMac.test(devices[i].getMAC());
+
+			// Only "real" interfaces for slaves needed
+			if (NotUsable == false) {
+				for (var j = 0; j < interfaces.length; j++) {
+					if (uci.get('network', interfaces[j]['.name'], 'proto') == 'bonding') {
+						var slaves = uci.get('network', interfaces[j]['.name'], 'slaves');
+						if (slaves != null) {
+							for (var k = 0; k < slaves.length; k++) {
+								if (devices[i].ifname == slaves[k] && interfaces[j]['.name'] != section_id) {
+									in_use = true;
+								}
+							}
+						}
+					}
+				}
+				if (in_use == false) {
+					rv.push(devices[i].ifname);
+				}
+			}
+		}
+
+	return rv.sort();
+	});
+}
+
+function validateEmpty(section, value) {
+	if (value) {
+		return true;
+	}
+	else {
+		return _('Expecting: non-empty value');
+	}
+}
+
+function updatePrimaries(section, value) {
+
+	var opt = this.map.lookupOption('slaves', section);
+	var selected_slaves = opt[0].formvalue(section);
+
+	var uielem = this.map.lookupOption('primary', section)[0].getUIElement(section);
+	uielem.clearChoices();
+
+	for (var i = 0; i < selected_slaves.length; i++) {
+		uielem.addChoices(selected_slaves[i], selected_slaves[i]);
+	}
+
+	return true;
+}
+
+function validate_arp_policy(section, value) {
+
+	var opt = this.map.lookupOption('link_monitoring', section);
+	var selected_link_monitoring = opt[0].formvalue(section);
+
+	var opt = this.map.lookupOption('bonding_policy', section);
+	var selected_policy = opt[0].formvalue(section);
+
+	if (selected_link_monitoring == 'arp') {
+		if (selected_policy == '802.3ad' || selected_policy == 'balance-tlb' || selected_policy == 'balance-alb') {
+			return _('ARP monitoring is not supported for the selected policy!');
+		}
+	}
+
+	return true;
+}
+
+function validate_arp_ip_targets(section, value) {
+
+	var opt = this.map.lookupOption('link_monitoring', section);
+	var selected_link_monitoring = opt[0].formvalue(section);
+
+	var opt = this.map.lookupOption('arp_ip_target', section);
+	var selected_arp_ip_targets = opt[0].formvalue(section);
+
+	var opt = this.map.lookupOption('bonding_policy', section);
+	var selected_policy = opt[0].formvalue(section);
+
+	if (selected_link_monitoring == 'arp' && selected_arp_ip_targets.length == 0) {
+		return _('You must select at least one ARP IP target if ARP monitoring is selected!');
+	}
+
+	return true;
+}
+
+function validate_primary_interface(section, value) {
+
+	var primary_valid = 'false';
+
+	var opt = this.map.lookupOption('bonding_policy', section);
+	var selected_policy = opt[0].formvalue(section);
+
+	var opt = this.map.lookupOption('slaves', section);
+	var selected_slaves = opt[0].formvalue(section);
+
+	var opt = this.map.lookupOption('primary', section);
+	var selected_primary = opt[0].formvalue(section);
+
+	if (selected_policy == 'active-backup' || selected_policy == 'balance-tlb' || selected_policy == 'balance-alb') {
+
+		for (var i = 0; i < selected_slaves.length; i++) {
```

Comment:
I believe you can shorten this code to something like this:

```
if (selected_slaves.filter(function(slave) { return slave == selected_primary }).length == 0)
    return _('You must select a primary interface which is included in selected slave interfaces!');
```

Reply:
OK, I made it following your suggestion.